# Code Datenbereinigung

Im nachfolgenden Code werden die gescrapten Daten für die Auswertung von Tobias Gerhard aufbereitet. Da die Daten nicht viele Verunreinigungen von sich aus vorweisen, werden zu Beginn noch Verunreinigungen von Hand eingefügt. Danach werden die einzelnen Verunreinigungen behoben, sowie die Tabelle um eine Spalte erweitert.

###  Import von notwendigen Librarys

In [67]:
import pandas as pd
import re

### Laden der gescrapten Daten

In [68]:
grammy_relevant_d = pd.read_csv("Suter_Sandro_src.csv")

### Übersicht über die gescrapten Daten verschaffen

In [69]:
grammy_relevant_d

Unnamed: 0  Unnamed: 1  Unnamed: 2  \
0            Record Of The Year        2020           0   
1            Record Of The Year        2019           0   
2            Record Of The Year        2018           0   
3            Record Of The Year        2017           0   
4            Record Of The Year        2016           0   
..                          ...         ...         ...   
113  Best Long Form Music Video        2004         106   
114  Best Long Form Music Video        2003         104   
115  Best Long Form Music Video        2002         103   
116  Best Long Form Music Video        2001         100   
117  Best Long Form Music Video        2000          99   

                       category  \
0            Record Of The Year   
1            Record Of The Year   
2            Record Of The Year   
3            Record Of The Year   
4            Record Of The Year   
..                          ...   
113  Best Long Form Music Video   
114  Best Long Form Music Video   
115  Best Long Form Music Video   
116  Best Long Form Music Video   
117  Best Long Form Music Video   

                                            Song titel  \
0                                  Everything I Wanted   
1                                              Bad Guy   
2                                      This Is America   
3                                            24K Magic   
4                                                Hello   
..                                                 ...   
113                                 Concert For George   
114                                             Legend   
115                               Westway To The World   
116  Recording The Producers - A Musical Romp With ...   
117  Gimme Some Truth - The Making Of John Lennon's...   

                                                artist artist2  
0                                        Billie Eilish     NaN  
1                                        Billie Eilish     NaN  
2                                     Childish Gambino     NaN  
3                                           Bruno Mars     NaN  
4                                                Adele     NaN  
..                                                 ...     ...  
113                                  (Various Artists)     NaN  
114                                        (Sam Cooke)     NaN  
115                                          The Clash     NaN  
116  Mel Brooks (Various Artists Including Nathan L...     NaN  
117                                      (John Lennon)     NaN  

[118 rows x 7 columns]

# Zusätzliche Verunreinigungen hinzufügen

### Jahr 2019 zu 20019 wechseln

Mit dieser Verunreinigung soll ein Fehler auf der Website simuliert werden. Das Jahr 2019 wäre in diesem Falle immer falsch als 20019 geschrieben. Entsprechend diesem Fehler kommt dies auch in unseren gescrapten Daten so vor.

In [70]:
grammy_relevant_d['Unnamed: 1'].replace(2019, 20019, inplace=True)
grammy_relevant_d

Unnamed: 0  Unnamed: 1  Unnamed: 2  \
0            Record Of The Year        2020           0   
1            Record Of The Year       20019           0   
2            Record Of The Year        2018           0   
3            Record Of The Year        2017           0   
4            Record Of The Year        2016           0   
..                          ...         ...         ...   
113  Best Long Form Music Video        2004         106   
114  Best Long Form Music Video        2003         104   
115  Best Long Form Music Video        2002         103   
116  Best Long Form Music Video        2001         100   
117  Best Long Form Music Video        2000          99   

                       category  \
0            Record Of The Year   
1            Record Of The Year   
2            Record Of The Year   
3            Record Of The Year   
4            Record Of The Year   
..                          ...   
113  Best Long Form Music Video   
114  Best Long Form Music Video   
115  Best Long Form Music Video   
116  Best Long Form Music Video   
117  Best Long Form Music Video   

                                            Song titel  \
0                                  Everything I Wanted   
1                                              Bad Guy   
2                                      This Is America   
3                                            24K Magic   
4                                                Hello   
..                                                 ...   
113                                 Concert For George   
114                                             Legend   
115                               Westway To The World   
116  Recording The Producers - A Musical Romp With ...   
117  Gimme Some Truth - The Making Of John Lennon's...   

                                                artist artist2  
0                                        Billie Eilish     NaN  
1                                        Billie Eilish     NaN  
2                                     Childish Gambino     NaN  
3                                           Bruno Mars     NaN  
4                                                Adele     NaN  
..                                                 ...     ...  
113                                  (Various Artists)     NaN  
114                                        (Sam Cooke)     NaN  
115                                          The Clash     NaN  
116  Mel Brooks (Various Artists Including Nathan L...     NaN  
117                                      (John Lennon)     NaN  

[118 rows x 7 columns]

### Duplikat einfügen

Auch hier wird ein Fehler auf der Website simuliert. Der Eintrag mit der Index-Nummer 50 kommt 2x vor, was einem Duplikat entspricht. Dies muss später entfernt werden.

In [71]:
grammy_relevant_d = grammy_relevant_d.append(grammy_relevant_d.iloc[50])
grammy_relevant_d

Unnamed: 0  Unnamed: 1  Unnamed: 2  \
0            Record Of The Year        2020           0   
1            Record Of The Year       20019           0   
2            Record Of The Year        2018           0   
3            Record Of The Year        2017           0   
4            Record Of The Year        2016           0   
..                          ...         ...         ...   
114  Best Long Form Music Video        2003         104   
115  Best Long Form Music Video        2002         103   
116  Best Long Form Music Video        2001         100   
117  Best Long Form Music Video        2000          99   
50            Album Of The Year        2012           1   

                       category  \
0            Record Of The Year   
1            Record Of The Year   
2            Record Of The Year   
3            Record Of The Year   
4            Record Of The Year   
..                          ...   
114  Best Long Form Music Video   
115  Best Long Form Music Video   
116  Best Long Form Music Video   
117  Best Long Form Music Video   
50            Album Of The Year   

                                            Song titel  \
0                                  Everything I Wanted   
1                                              Bad Guy   
2                                      This Is America   
3                                            24K Magic   
4                                                Hello   
..                                                 ...   
114                                             Legend   
115                               Westway To The World   
116  Recording The Producers - A Musical Romp With ...   
117  Gimme Some Truth - The Making Of John Lennon's...   
50                                               Babel   

                                                artist artist2  
0                                        Billie Eilish     NaN  
1                                        Billie Eilish     NaN  
2                                     Childish Gambino     NaN  
3                                           Bruno Mars     NaN  
4                                                Adele     NaN  
..                                                 ...     ...  
114                                        (Sam Cooke)     NaN  
115                                          The Clash     NaN  
116  Mel Brooks (Various Artists Including Nathan L...     NaN  
117                                      (John Lennon)     NaN  
50                                      Mumford & Sons     NaN  

[119 rows x 7 columns]

### Verunreinigte csv-Datei abspeichern

In [72]:
grammy_relevant_d.to_csv("Suter_Sandro_dirty.csv", index=False,header=True)

# Verunreinigungen aus dem gescrapten DataFrame beseitigen

### Verunreinigte csv-Datei laden

In [73]:
grammy_relevant_dirty = pd.read_csv("Suter_Sandro_dirty.csv")
grammy_relevant_dirty

Unnamed: 0  Unnamed: 1  Unnamed: 2  \
0            Record Of The Year        2020           0   
1            Record Of The Year       20019           0   
2            Record Of The Year        2018           0   
3            Record Of The Year        2017           0   
4            Record Of The Year        2016           0   
..                          ...         ...         ...   
114  Best Long Form Music Video        2003         104   
115  Best Long Form Music Video        2002         103   
116  Best Long Form Music Video        2001         100   
117  Best Long Form Music Video        2000          99   
118           Album Of The Year        2012           1   

                       category  \
0            Record Of The Year   
1            Record Of The Year   
2            Record Of The Year   
3            Record Of The Year   
4            Record Of The Year   
..                          ...   
114  Best Long Form Music Video   
115  Best Long Form Music Video   
116  Best Long Form Music Video   
117  Best Long Form Music Video   
118           Album Of The Year   

                                            Song titel  \
0                                  Everything I Wanted   
1                                              Bad Guy   
2                                      This Is America   
3                                            24K Magic   
4                                                Hello   
..                                                 ...   
114                                             Legend   
115                               Westway To The World   
116  Recording The Producers - A Musical Romp With ...   
117  Gimme Some Truth - The Making Of John Lennon's...   
118                                              Babel   

                                                artist artist2  
0                                        Billie Eilish     NaN  
1                                        Billie Eilish     NaN  
2                                     Childish Gambino     NaN  
3                                           Bruno Mars     NaN  
4                                                Adele     NaN  
..                                                 ...     ...  
114                                        (Sam Cooke)     NaN  
115                                          The Clash     NaN  
116  Mel Brooks (Various Artists Including Nathan L...     NaN  
117                                      (John Lennon)     NaN  
118                                     Mumford & Sons     NaN  

[119 rows x 7 columns]

### Nicht notwendige Spalten entfernen

Beim Scraping wurde zuerst für jedes Jahr ein DataFrame mit den relevanten Kategorien erstellt.

Da beim zusammenführen zu dem finalen DataFrame (grammy_relevant.csv) wurde der concat-Befehl aus der Bibliothek "Pandas" verwendet. Dieser fügt die DataFrames über die sogenannten Keys zusammen. Als Keys wurden die relevanten Kategorien gewählt.

Dies führte dazu, dass die Grammy-Kategorien 2x im finalen DataFrame auftauchen. Weiter wurde auch die Index-Nummer der einzelnen DataFrames im finalen DataFrame abgebildet. Somit sind die Spalten "category" und "Unnamed: 2" nicht mehr von nöten und können gelöscht werden.

Dies geschieht über die _.drop()_-Methode des Pandas-DataFrames. Mit _"columns='...' "_ wird die Spalte gewählt, welche gelöscht werden soll. Durch den Befehl _"inplace=True"_ wird angegeben, dass die Methode direkt auf den Originaldaten ausgeführt werden soll.

In [74]:
grammy_relevant_dirty.drop(columns='Unnamed: 2', inplace=True)
grammy_relevant_dirty.drop(columns='category', inplace=True)
grammy_relevant_dirty

Unnamed: 0  Unnamed: 1  \
0            Record Of The Year        2020   
1            Record Of The Year       20019   
2            Record Of The Year        2018   
3            Record Of The Year        2017   
4            Record Of The Year        2016   
..                          ...         ...   
114  Best Long Form Music Video        2003   
115  Best Long Form Music Video        2002   
116  Best Long Form Music Video        2001   
117  Best Long Form Music Video        2000   
118           Album Of The Year        2012   

                                            Song titel  \
0                                  Everything I Wanted   
1                                              Bad Guy   
2                                      This Is America   
3                                            24K Magic   
4                                                Hello   
..                                                 ...   
114                                             Legend   
115                               Westway To The World   
116  Recording The Producers - A Musical Romp With ...   
117  Gimme Some Truth - The Making Of John Lennon's...   
118                                              Babel   

                                                artist artist2  
0                                        Billie Eilish     NaN  
1                                        Billie Eilish     NaN  
2                                     Childish Gambino     NaN  
3                                           Bruno Mars     NaN  
4                                                Adele     NaN  
..                                                 ...     ...  
114                                        (Sam Cooke)     NaN  
115                                          The Clash     NaN  
116  Mel Brooks (Various Artists Including Nathan L...     NaN  
117                                      (John Lennon)     NaN  
118                                     Mumford & Sons     NaN  

[119 rows x 5 columns]

### Benötigte Spalten bezeichnen

Durch das Zusammenführen der verschiedenen DataFrames je Jahr sind Spalten ohne Namen entstanden. Diese müssen nun benennt werden.

Dies wird über die _.rename()_-Methode gemacht. Durch die Übergabe eines Dictionaries, in welchem der _Key= 'Bestehender Spaltenname'_ und der _Value= 'Gewünschter Spaltenname'_ ist, können die Spaltennamen angepasst werden. Auch hier wird wieder mit der Einstellung _inplace=True_ gearbeitet, damit direkt die Originaldaten angepasst werden. 

In [75]:
grammy_relevant_dirty.rename(columns={'Unnamed: 0': 'Kategorie', 'Unnamed: 1': 'Jahr', 'Song titel': 'Winner','artist':'Künstler/in'}, inplace=True)

In [76]:
grammy_relevant_dirty

Kategorie   Jahr  \
0            Record Of The Year   2020   
1            Record Of The Year  20019   
2            Record Of The Year   2018   
3            Record Of The Year   2017   
4            Record Of The Year   2016   
..                          ...    ...   
114  Best Long Form Music Video   2003   
115  Best Long Form Music Video   2002   
116  Best Long Form Music Video   2001   
117  Best Long Form Music Video   2000   
118           Album Of The Year   2012   

                                                Winner  \
0                                  Everything I Wanted   
1                                              Bad Guy   
2                                      This Is America   
3                                            24K Magic   
4                                                Hello   
..                                                 ...   
114                                             Legend   
115                               Westway To The World   
116  Recording The Producers - A Musical Romp With ...   
117  Gimme Some Truth - The Making Of John Lennon's...   
118                                              Babel   

                                           Künstler/in artist2  
0                                        Billie Eilish     NaN  
1                                        Billie Eilish     NaN  
2                                     Childish Gambino     NaN  
3                                           Bruno Mars     NaN  
4                                                Adele     NaN  
..                                                 ...     ...  
114                                        (Sam Cooke)     NaN  
115                                          The Clash     NaN  
116  Mel Brooks (Various Artists Including Nathan L...     NaN  
117                                      (John Lennon)     NaN  
118                                     Mumford & Sons     NaN  

[119 rows x 5 columns]

### Jahreszahlen von 20019 zu 2019 wechseln

Die von Hand eingeführte Verunreinigung des Jahres 2019 zu _20019_ wird mit der _.replace()_-Methode behoben. Die Spalte _'Jahr'_ wird dabei auf den Ausdruck _20019_ untersucht und dieser wird immer durch _2019_ ersetzt. Auch diese Anpassung geschieht inplace.

In [77]:
grammy_relevant_dirty['Jahr'].replace(20019,2019, inplace=True)
grammy_relevant_dirty

Kategorie  Jahr  \
0            Record Of The Year  2020   
1            Record Of The Year  2019   
2            Record Of The Year  2018   
3            Record Of The Year  2017   
4            Record Of The Year  2016   
..                          ...   ...   
114  Best Long Form Music Video  2003   
115  Best Long Form Music Video  2002   
116  Best Long Form Music Video  2001   
117  Best Long Form Music Video  2000   
118           Album Of The Year  2012   

                                                Winner  \
0                                  Everything I Wanted   
1                                              Bad Guy   
2                                      This Is America   
3                                            24K Magic   
4                                                Hello   
..                                                 ...   
114                                             Legend   
115                               Westway To The World   
116  Recording The Producers - A Musical Romp With ...   
117  Gimme Some Truth - The Making Of John Lennon's...   
118                                              Babel   

                                           Künstler/in artist2  
0                                        Billie Eilish     NaN  
1                                        Billie Eilish     NaN  
2                                     Childish Gambino     NaN  
3                                           Bruno Mars     NaN  
4                                                Adele     NaN  
..                                                 ...     ...  
114                                        (Sam Cooke)     NaN  
115                                          The Clash     NaN  
116  Mel Brooks (Various Artists Including Nathan L...     NaN  
117                                      (John Lennon)     NaN  
118                                     Mumford & Sons     NaN  

[119 rows x 5 columns]

### Duplikat entfernen

Das am Ende des DataFrames eingefügte Duplikat muss nun mittels einer Methode entfernt werden. Bei der Durchsicht des DataFrames kann festgestellt werden, dass am Schluss die Index-Nummer 50 nochmals vorkommt. Dies lässt vermuten, dass es sich um ein Duplikat handelt. 

Daher werden zuerst die Indexnummern 50 und 118 angeschaut, Ziel ist es zu vergleichen ob es ein Dupplikat ist:

In [78]:
print(grammy_relevant_dirty.iloc[50])
print("*"*10)
print(grammy_relevant_dirty.iloc[118])

Kategorie      Album Of The Year
Jahr                        2012
Winner                     Babel
Künstler/in       Mumford & Sons
artist2                      NaN
Name: 50, dtype: object
**********
Kategorie      Album Of The Year
Jahr                        2012
Winner                     Babel
Künstler/in       Mumford & Sons
artist2                      NaN
Name: 118, dtype: object


Es lässt sich unschwer erkennen, dass es sich um ein Duplikat handelt, welches entfernt werden muss (beide sind Album Of The Year, 2012).

Dazu wird die Methode _.drop_duplicates()_ verwendet. Diese geht das gesamte DataFrame durch und entfernt allfällige Duplikate:

In [79]:
grammy_relevant_dirty = grammy_relevant_dirty.drop_duplicates()
grammy_relevant_dirty

Kategorie  Jahr  \
0            Record Of The Year  2020   
1            Record Of The Year  2019   
2            Record Of The Year  2018   
3            Record Of The Year  2017   
4            Record Of The Year  2016   
..                          ...   ...   
113  Best Long Form Music Video  2004   
114  Best Long Form Music Video  2003   
115  Best Long Form Music Video  2002   
116  Best Long Form Music Video  2001   
117  Best Long Form Music Video  2000   

                                                Winner  \
0                                  Everything I Wanted   
1                                              Bad Guy   
2                                      This Is America   
3                                            24K Magic   
4                                                Hello   
..                                                 ...   
113                                 Concert For George   
114                                             Legend   
115                               Westway To The World   
116  Recording The Producers - A Musical Romp With ...   
117  Gimme Some Truth - The Making Of John Lennon's...   

                                           Künstler/in artist2  
0                                        Billie Eilish     NaN  
1                                        Billie Eilish     NaN  
2                                     Childish Gambino     NaN  
3                                           Bruno Mars     NaN  
4                                                Adele     NaN  
..                                                 ...     ...  
113                                  (Various Artists)     NaN  
114                                        (Sam Cooke)     NaN  
115                                          The Clash     NaN  
116  Mel Brooks (Various Artists Including Nathan L...     NaN  
117                                      (John Lennon)     NaN  

[118 rows x 5 columns]

### Entfernen von "()" aus der Spalte "Künstler/in"

In der Spalte "Künstler/in" gibt es mehrere Einträge, welche fälschlicherweise in einer Klammer geschrieben werden. Als Beispiel wird hier die Index-Nr. 115 gezeigt:

*Best Long Form Music Video	2003	Legend	**(Sam Cooke)**	,No Interpreten,*

Sam Cooke sollte hier nicht in den Klammern stehen. Besonders zu beachten ist, dass es Zeilen gibt, bei denen die Klammern Sinn ergeben. Bei diesen sollen die Klammern weiterhin bestehen. Als Beispiel wird die Index-Nr. 116 genommen:

*Best Long Form Music Video	2001	Recording The Producers - A Musical Romp With ...	**Mel Brooks (Various Artists Including Nathan L...**	,No Interpreten,*

Hierbei ist der Text innerhalb der Klammer eine Anmerkung und soll so bestehen bleiben.

Um ein Überblick zu bekommen, wo dass es alles Klammern gibt, die entfernt werden müssen und wo es solche hat, die bestehen bleiben sollen, wird zuerst die _.unique()_-Methode angewandt. Diese zeigt sämtliche Werte, die in der Spalte _'Künstler/in'_ vorkommt.

In [80]:
print(grammy_relevant_dirty['Künstler/in'].unique())

['Billie Eilish' 'Childish Gambino' 'Bruno Mars' 'Adele'
 'Mark Ronson Featuring Bruno Mars' 'Sam Smith'
 'Daft Punk Featuring Pharrell Williams & Nile Rodgers'
 'Gotye Featuring Kimbra' 'Lady Antebellum' 'Kings Of Leon'
 'Robert Plant & Alison Krauss' 'Amy Winehouse' 'Dixie Chicks' 'Green Day'
 'Ray Charles & Norah Jones' 'Coldplay' 'Norah Jones' 'U2' nan
 'Taylor Swift' 'Kacey Musgraves' 'Beck' 'Daft Punk' 'Mumford & Sons'
 'Arcade Fire' 'Herbie Hancock' 'Ray Charles & Various Artists' 'OutKast'
 '(Various Artists)' 'Steely Dan' 'Beyoncé, Blue Ivy & WizKid'
 'Lil Nas X & Billy Ray Cyrus' 'Kendrick Lamar' 'Beyoncé'
 'Taylor Swift Featuring Kendrick Lamar' 'Pharrell Williams'
 'Justin Timberlake Featuring Jay Z' 'Rihanna Featuring Calvin Harris'
 'Lady Gaga' 'The Black Eyed Peas' 'Weezer' 'Johnny Cash' 'OK Go'
 'Missy Elliott Featuring Ciara & Fat Man Scoop' 'Eminem'
 'Fatboy Slim Featuring Bootsy Collins' 'Foo Fighters'
 'Mumford & Sons, Edward Sharpe & The Magnetic Zeros & Old Crow M

Es lässt sich schnell erkennen, dass obiges Beispiel (Zeile 116) die einzige Zeile ist, die berechtigte Klammern hat. Somit müssen sämtliche Klammern aus der Spalte _'Künstler/in'_ entfernt werden, bis auf diejenigen der Zeile 116.

Zuerst werden sämtliche Klammern am Anfang und am Ende jedes Strings mittels regex entfernt. Dazu wird mit der Methode _.loc[]_ die Spalte _'Künstler/in'_ aufgerufen und mit der Methode _.replace()_ die Klammern durch _''_ (nichts) ersetzt.

Innerhalb der _replace()_-Methode wird ein Regex angewendet. Dies wird dann auch mittels _regex=True_ definiert. Der angewendete Regex-Befehl ist:

**^\(|)$**

Durch das Zeichen **^** wird zuerst am Anfang des Strings nach **(** gesucht. Durch das Zeichen **|** wird dann zur nächsten Expression gesprungen. Diese sucht nach **)** am Schluss des Stings.

Zusammenfassend wird durch den Regex **^\(|\)$** die erste und letzte Klammer eines Stings ausgegeben. Beispiel:

**(**Sam Cook**)**

Mel Brooks (Various Artists Including Nathan Lane & Matthew Broderick**)**

**(**John Lenon**)**

In [81]:
grammy_relevant_dirty.loc[:'Künstler/in'].replace('^\(|\)$', '', regex=True, inplace=True)


Auf der Zeile 116 ist nun eine Klammer zu viel entfernt. Daher wird dieser Eintrag wieder von Hand korrigiert. Auch hier wird mittels der _.loc_-Methode auf den Eintrag zugegeriffen und dieser mittels _=_ manuell angepasst.

In [82]:
grammy_relevant_dirty.loc[116,'Künstler/in']='Mel Brooks (Various Artists Including Nathan Lane & Matthew Broderick)'

/home/student/venv/pds01-jpy-3.8/lib/python3.8/site-packages/pandas/core/indexing.py:1720: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


In [83]:
grammy_relevant_dirty['Künstler/in'][116]

'Mel Brooks (Various Artists Including Nathan Lane & Matthew Broderick)'

# Enrichmentsteps

### Spalten "artist2" & "Künstler/in" zu "_Künstler/in umfassend" zusammenlegen

Die Spalte "artist2" wurde beim Scraping aufgrund der Beschaffenheit der Webseite der Grammy-Awards erstellt. Die Namen der Künstlerinnen und Künstler, welche jeweils gewonnen haben, waren dort nicht immer im gleichen HTML-Element.

Nun gilt es eine neue Spalte _'_Künstler/in umfassend'_ zu erstellen, welche keine NaN Werte mehr enthält und jeweils den oder die Künstlerin enthaltet, welche die Kategorie im entsprechenden Jahr gewonnen hat.

Dazu wird zuerst geschaut wie viele NaN es hat:

In [84]:
print("Anzahl NaN in 'Künstler/in' vorher: ", grammy_relevant_dirty['Künstler/in'].isna().sum())
print("Anzahl NaN in 'artist2' vorher: ", grammy_relevant_dirty['artist2'].isna().sum())

Anzahl NaN in 'Künstler/in' vorher:  42
Anzahl NaN in 'artist2' vorher:  97


Als nächstes werden diese NaN-Werte durch den String _',No Interpreten,'_ ersetzt, und es wird Kontrolliert ob es keine NaN-Werte mehr gibt:

In [85]:
grammy_relevant_dirty['artist2'] = grammy_relevant_dirty['artist2'].fillna(',No Interpreten,')
grammy_relevant_dirty['Künstler/in'] = grammy_relevant_dirty['Künstler/in'].fillna(',No Interpreten,')

print("Anzahl NaN in 'artist2' nachher: ", grammy_relevant_dirty['artist2'].isna().sum())
print("Anzahl NaN in 'Künstler/in' nachher: ", grammy_relevant_dirty['Künstler/in'].isna().sum())

Anzahl NaN in 'artist2' nachher:  0
Anzahl NaN in 'Künstler/in' nachher:  0


<ipython-input-85-57dbf4c4bf88>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  grammy_relevant_dirty['artist2'] = grammy_relevant_dirty['artist2'].fillna(',No Interpreten,')
<ipython-input-85-57dbf4c4bf88>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  grammy_relevant_dirty['Künstler/in'] = grammy_relevant_dirty['Künstler/in'].fillna(',No Interpreten,')


Anschliessend werden die beiden Spalten _'Künstler/in'_ und '_artist2_' zu einer neuen Spalte 'Künstler/in umfassend' zusammengelegt. Dies geschieht einfach mit dem + Operator.

Wichtig zu beachten ist, dass die neue Spalte _'Künstler/in umfassend'_ mit einem "_" gekennzeichnet ist, damit Andere Betrachter sofort sehen, dass es sich um eine angereichtere Spalte handelt:

In [86]:
grammy_relevant_dirty['_Künstler/in umfassend'] = grammy_relevant_dirty['Künstler/in'] + grammy_relevant_dirty['artist2']

grammy_relevant_dirty

<ipython-input-86-36fe58ae5ba2>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  grammy_relevant_dirty['_Künstler/in umfassend'] = grammy_relevant_dirty['Künstler/in'] + grammy_relevant_dirty['artist2']


Kategorie  Jahr  \
0            Record Of The Year  2020   
1            Record Of The Year  2019   
2            Record Of The Year  2018   
3            Record Of The Year  2017   
4            Record Of The Year  2016   
..                          ...   ...   
113  Best Long Form Music Video  2004   
114  Best Long Form Music Video  2003   
115  Best Long Form Music Video  2002   
116  Best Long Form Music Video  2001   
117  Best Long Form Music Video  2000   

                                                Winner  \
0                                  Everything I Wanted   
1                                              Bad Guy   
2                                      This Is America   
3                                            24K Magic   
4                                                Hello   
..                                                 ...   
113                                 Concert For George   
114                                             Legend   
115                               Westway To The World   
116  Recording The Producers - A Musical Romp With ...   
117  Gimme Some Truth - The Making Of John Lennon's...   

                                           Künstler/in           artist2  \
0                                        Billie Eilish  ,No Interpreten,   
1                                        Billie Eilish  ,No Interpreten,   
2                                     Childish Gambino  ,No Interpreten,   
3                                           Bruno Mars  ,No Interpreten,   
4                                                Adele  ,No Interpreten,   
..                                                 ...               ...   
113                                    Various Artists  ,No Interpreten,   
114                                          Sam Cooke  ,No Interpreten,   
115                                          The Clash  ,No Interpreten,   
116  Mel Brooks (Various Artists Including Nathan L...  ,No Interpreten,   
117                                        John Lennon  ,No Interpreten,   

                                _Künstler/in umfassend  
0                        Billie Eilish,No Interpreten,  
1                        Billie Eilish,No Interpreten,  
2                     Childish Gambino,No Interpreten,  
3                           Bruno Mars,No Interpreten,  
4                                Adele,No Interpreten,  
..                                                 ...  
113                    Various Artists,No Interpreten,  
114                          Sam Cooke,No Interpreten,  
115                          The Clash,No Interpreten,  
116  Mel Brooks (Various Artists Including Nathan L...  
117                        John Lennon,No Interpreten,  

[118 rows x 6 columns]

Bei der Kontrolle der angereicherten Spalte kann nun festgestellt werden, dass immer am Ende des Strings der Zusatz ',No Interpreten,' angefügt wurde. Dies geschieht durch den '+' Operator. Dieser Zusatz muss nun noch in jedem String entfernt werden.

Auch dies wird mit einem Regex-Befehl in der _.replace()_-Methode gemacht. Der Regex:

**(,\w* \w*,)**

sucht nach 2 Wörtern innerhalb von zwei Kommas. Die gefundenen Teil-Strings werden dann durch _''_ (nichts) ersetzt.

In [87]:
grammy_relevant_dirty['_Künstler/in umfassend'].replace('(,\w* \w*,)', '', regex=True, inplace= True)
grammy_relevant_dirty[0:50]

/home/student/venv/pds01-jpy-3.8/lib/python3.8/site-packages/pandas/core/series.py:4506: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().replace(


Kategorie  Jahr                                    Winner  \
0   Record Of The Year  2020                       Everything I Wanted   
1   Record Of The Year  2019                                   Bad Guy   
2   Record Of The Year  2018                           This Is America   
3   Record Of The Year  2017                                 24K Magic   
4   Record Of The Year  2016                                     Hello   
5   Record Of The Year  2015                               Uptown Funk   
6   Record Of The Year  2014           Stay With Me (Darkchild Version   
7   Record Of The Year  2013                                 Get Lucky   
8   Record Of The Year  2012              Somebody That I Used To Know   
9   Record Of The Year  2011                       Rolling In The Deep   
10  Record Of The Year  2010                              Need You Now   
11  Record Of The Year  2009                              Use Somebody   
12  Record Of The Year  2008                    Please Read The Letter   
13  Record Of The Year  2007                                     Rehab   
14  Record Of The Year  2006                    Not Ready To Make Nice   
15  Record Of The Year  2005                Boulevard Of Broken Dreams   
16  Record Of The Year  2004                          Here We Go Again   
17  Record Of The Year  2003                                    Clocks   
18  Record Of The Year  2002                            Don't Know Why   
19  Record Of The Year  2001                                   Walk On   
20  Record Of The Year  2000                             Beautiful Day   
21  Artist Of The Year  2020                       Megan Thee Stallion   
22  Artist Of The Year  2019                             Billie Eilish   
23  Artist Of The Year  2018                                  Dua Lipa   
24  Artist Of The Year  2017                              Alessia Cara   
25  Artist Of The Year  2016                         Chance The Rapper   
26  Artist Of The Year  2015                            Meghan Trainor   
27  Artist Of The Year  2014                                 Sam Smith   
28  Artist Of The Year  2013                   Macklemore & Ryan Lewis   
29  Artist Of The Year  2012                                      Fun.   
30  Artist Of The Year  2011                                  Bon Iver   
31  Artist Of The Year  2010                        Esperanza Spalding   
32  Artist Of The Year  2009                            Zac Brown Band   
33  Artist Of The Year  2008                                     Adele   
34  Artist Of The Year  2007                             Amy Winehouse   
35  Artist Of The Year  2006                          Carrie Underwood   
36  Artist Of The Year  2005                               John Legend   
37  Artist Of The Year  2004                                  Maroon 5   
38  Artist Of The Year  2003                               Evanescence   
39  Artist Of The Year  2002                               Norah Jones   
40  Artist Of The Year  2001                               Alicia Keys   
41  Artist Of The Year  2000                              Shelby Lynne   
42   Album Of The Year  2020                                  Folklore   
43   Album Of The Year  2019  When We All Fall Asleep, Where Do We Go?   
44   Album Of The Year  2018                               Golden Hour   
45   Album Of The Year  2017                                 24K Magic   
46   Album Of The Year  2016                                        25   
47   Album Of The Year  2015                                      1989   
48   Album Of The Year  2014                             Morning Phase   
49   Album Of The Year  2013                    Random Access Memories   

                                          Künstler/in           artist2  \
0                                       Billie Eilish  ,No Interpreten,   
1                                       Billie Eilish  ,No Interpreten,   
2            

### Kategorie: "Artist Of The Year"

Die Spalte Winner bezeichnet immer den Winner-Song. Da bei der Kategorie "Artist Of The Year" der Gewinner nicht ein Song, sondern ein Künstler ist, müssen diese Einträge noch von der Spalte "Winner" in die Spalte "_Künstler/in umfassend" eingetragen werden

In [88]:
grammy_relevant_dirty.loc[21:41,'_Künstler/in umfassend'] = grammy_relevant_dirty.loc[21:41,'Winner']

/home/student/venv/pds01-jpy-3.8/lib/python3.8/site-packages/pandas/core/indexing.py:1676: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


In [89]:
grammy_relevant_dirty[15:50]

Kategorie  Jahr                                    Winner  \
15  Record Of The Year  2005                Boulevard Of Broken Dreams   
16  Record Of The Year  2004                          Here We Go Again   
17  Record Of The Year  2003                                    Clocks   
18  Record Of The Year  2002                            Don't Know Why   
19  Record Of The Year  2001                                   Walk On   
20  Record Of The Year  2000                             Beautiful Day   
21  Artist Of The Year  2020                       Megan Thee Stallion   
22  Artist Of The Year  2019                             Billie Eilish   
23  Artist Of The Year  2018                                  Dua Lipa   
24  Artist Of The Year  2017                              Alessia Cara   
25  Artist Of The Year  2016                         Chance The Rapper   
26  Artist Of The Year  2015                            Meghan Trainor   
27  Artist Of The Year  2014                                 Sam Smith   
28  Artist Of The Year  2013                   Macklemore & Ryan Lewis   
29  Artist Of The Year  2012                                      Fun.   
30  Artist Of The Year  2011                                  Bon Iver   
31  Artist Of The Year  2010                        Esperanza Spalding   
32  Artist Of The Year  2009                            Zac Brown Band   
33  Artist Of The Year  2008                                     Adele   
34  Artist Of The Year  2007                             Amy Winehouse   
35  Artist Of The Year  2006                          Carrie Underwood   
36  Artist Of The Year  2005                               John Legend   
37  Artist Of The Year  2004                                  Maroon 5   
38  Artist Of The Year  2003                               Evanescence   
39  Artist Of The Year  2002                               Norah Jones   
40  Artist Of The Year  2001                               Alicia Keys   
41  Artist Of The Year  2000                              Shelby Lynne   
42   Album Of The Year  2020                                  Folklore   
43   Album Of The Year  2019  When We All Fall Asleep, Where Do We Go?   
44   Album Of The Year  2018                               Golden Hour   
45   Album Of The Year  2017                                 24K Magic   
46   Album Of The Year  2016                                        25   
47   Album Of The Year  2015                                      1989   
48   Album Of The Year  2014                             Morning Phase   
49   Album Of The Year  2013                    Random Access Memories   

                  Künstler/in           artist2     _Künstler/in umfassend  
15                  Green Day  ,No Interpreten,                  Green Day  
16  Ray Charles & Norah Jones  ,No Interpreten,  Ray Charles & Norah Jones  
17                   Coldplay  ,No Interpreten,                   Coldplay  
18                Norah Jones  ,No Interpreten,                Norah Jones  
19                         U2  ,No Interpreten,                         U2  
20                         U2  ,No Interpreten,                         U2  
21           ,No Interpreten,  ,No Interpreten,        Megan Thee Stallion  
22           ,No Interpreten,  ,No Interpreten,              Billie Eilish  
23           ,No Interpreten,  ,No Interpreten,                   Dua Lipa  
24           ,No Interpreten,  ,No Interpreten,               Alessia Cara  
25           ,No Interpreten,  ,No Interpreten,          Chance The Rapper  
26           ,No Interpreten,  ,No Interpreten,             Meghan Trainor  
27           ,No Interpreten,  ,No Interpreten,                  Sam Smith  
28           ,No Interpreten,  ,No Interpreten,    Macklemore & Ryan Lewis  
29           ,No Interpreten,  ,No Interpreten,                       Fun.  
30           ,No Interpreten,  ,No Interpreten,                   Bon Iver  
31           ,No Interpreten,  ,No Inte

# Abspeichern als csv-Datei

In [90]:
grammy_relevant_dirty.to_csv("Suter_Sandro_stage.csv",index=False,header=True)

# Database Connection

Nachfolgend wird die Database Connection für die MariaDB erstellt. Dieser Code wurde von Student C erarbeitet und den Studenten A und B zur Verfügung gestellt. Die letzteren beiden konnten diesen lediglich kopieren und die nachfolgenden Attribute anpassen:

- User
- Password
- Table
- Path

In [91]:
!pip install mariadb

In [92]:
import mariadb

In [93]:
#Verbindung aufbauen
#PW in Plain-Text ist natürlich nicht optimal... aber irgendwo hätte ich es sowieso aufschreiben müssen, deshalb
#habe ich darauf verzichtet, das PW zu hashen
try:
    conn = mariadb.connect(
      user="sandro",
      password="Frosch12",
      host="localhost",
      port=3306,
      autocommit=True) #So erübrigt sich conn.commit()        
except mariadb.Error as e:
    print(f"Error connecting to MariaDB Platform: {e}")
    sys.exit(1)

print("Verbindung erfolgreich aufgebaut.")

Verbindung erfolgreich aufgebaut.


In [94]:
#Cursor kreieren, über welchen die DB erreicht wird.
cur = conn.cursor()

In [95]:
#Nun können wir via Cursor SQL-Queries schreiben. Hier erstellen wir die Datenbank. Die SQL-Queries sollten
#selbsterklärend sein.
database_name = "CIP"

cur.execute("DROP DATABASE IF EXISTS " + database_name + ";") 
cur.execute("CREATE DATABASE " + database_name + ";")
cur.execute("USE " + database_name + ";")

In [96]:
db_table_name = "grammys"
csv_file_name_path = "/home/student/Cloud/Owncloud/SyncVM/CIP_FS21/Pycharm/CIP_Groupwork/TEST/Suter_Sandro_stage.csv"
field_terminator = ","
line_terminator = "\n"
lines_to_be_igored = "1" #so wird die 1. Line (Column-Names) ignoriert

In [97]:
#Table in der DB erstellen mit den entsprechenden Attributen
cur.execute("CREATE TABLE IF NOT EXISTS " + db_table_name + \
            "("
            "Kategorie varchar(255),"
            "Jahr varchar(255),"
            "Winner varchar(255),"
            "Künstler_in varchar(255),"
            "artist2 varchar(255),"
            "_Künstler_in_umfassend varchar(255)"
            ");"
            )

In [98]:
#Query zum Einlesen der Daten
query_insert = "LOAD DATA LOCAL INFILE '" + csv_file_name_path + \
                "' INTO TABLE " + db_table_name + \
                " FIELDS TERMINATED BY '" + field_terminator + \
                "' LINES TERMINATED BY '" + line_terminator + \
                "' IGNORE " + lines_to_be_igored + " LINES;"

#Ausführung des Queries bzw. tatsächlicher Import der Daten
cur.execute(query_insert)

In [99]:
#Üblicherweise würden wir hier noch die Connection schliessen, machen wir aber in unserem Fall nicht
#(aus Gründen der Bequemlichkeit, ansonsten müssten wir die Connection immer wieder aufbauen, zum Testen ist das nervig):
#conn.close()

# Lessons learned

Die Arbeit wurde von mir persönlich als sehr interessant und lehrreich empfunden. Die Durcharbeitung eines gesamten ETL-Prozesses anhand eines realen Beispiels zeigte gut auf, welche Schwierigkeiten und Herausforderungen auftauchen können.

- Mir persönlich ist vorallem die Wichtigkeit der Vorbereitung geblieben. Es hätte sich gelohnt, mehr Zeit in die Findung und Definition der Forschungsfrage zu investieren. Je genauer die Forschungsfrage definiert ist, desto genauer können die dafür benötigten Daten definiert und gesucht werden. Dies hätte besonders die Beantwortung der Fragen vereinfacht.


- Weiter habe ich festgestellt, dass die Wahl der Entwicklungsumgebung ein grossen Einfluss auf die Arbeit hat. PyCharm ist sehr gut geeignet für komplexere, zusammenhängende Codes, bei welchen die grafische Darstellung der Daten nicht im Vordergrund steht. Jupyter Notebook ist sehr gut geeingent für einfachere Codes, bei welchen einzelne Zeilen ausgegeben werden sollen und die grafische Darstellung wichtig ist.


- Besonders spannend war es zu sehen wie Zeitintensiv die Datenbereinigung ist. Obwohl die verarbeiteten Daten eher sauber waren, benötigte ich am meisten Zeit für diesen Schritt. Es war aus meiner Sicht ein sehr guter Ausblick auf die zukünftige Tätigkeit als DataScientist.


- Als letzte grosse "Lesson learned" war es spannend den handwerklichen Umgang mit einem HTML-Code. Im bisherigen Studium hatten, wir dies eher aus der theoretischen Sicht betrachtet. Wie der Umgang in der Praxis ist war sehr spannend und lehrreich.


Dies sind meine 4 grossen "Lessons learned". Es gab natürlich noch viele kleine weiteren wie der Umgang mit den einzelnen Python Bibliotheken Pandas und BeautifulSoup oder die Nutzerfreundlichkeit von Python. Abschliessend wurde nochmals bewusst, welch grosse Möglichkeiten mit Webscraping und dem beherrschen von Python entstehen. Dies ergab sicherlich nochmals einen grossen Motivationsschub diese Skills zu vertiefen.